src/irmin-pack/unix/io.ml

```diff
@@ -211,6 +211,23 @@ module Unix = struct
     | Errors.Closed -> Error `Closed
     | Unix.Unix_error (e, s1, s2) -> Error (`Io_misc (e, s1, s2))
 
+  let page_size = 4096
+
+  let read_all_to_string t =
+    let buf = Buffer.create 0 in
+    let len = page_size in
+    let bytes = Bytes.create len in
+    let rec aux ~off =
+      let nread = Util.really_read t.fd off len bytes in
```

Comment:
you still have a data race here if the RW instance update the file between two short reads, don't you?

Reply:
Two thoughts:

1. I don't think a loop is needed here since `Util.really_read` also loops until `len` is read. But --
2. Regardless of location of the loop, it does seem to me that a race can happen if you need multiple reads. What is the best (or even a good) way to protect against this? The simplest I can think of is to check last updated timestamp before each read and start over if it changes, but maybe there is some other trick?

Reply:
Here is how my patch functions:
- I've added a `read_all_to_string` function that just reads a file. It doesn't guarantee that the full content of the file is read atomically. It only guarantees that pages are read atomically (i.e. a page never spans on two iterations of the loop)
- I've changed the `read` function in the control file to rely on this new `read_all_to_string` function. Since control files are always contained within a page, that read is atomic. The assertion ensures so.

Following your reviews I've changed my code in the following manner:
- In `io_intf.ml` I've detailed the guarantees offered by `read_all_to_string`
- In `read_all_to_string` I am now directly using `Syscalls.pread` in order to make it more obvious that the pages are read atomically. The loop in `really_read` only exists for situations where `len` is greater than OCaml's `UNIX_BUFFER_SIZE` (https://github.com/ocaml/ocaml/blob/trunk/otherlibs/unix/unixsupport.h#L107, https://github.com/mirage/index/blob/main/src/unix/pread.c#L18)
- In `control_file.ml` I've documented and improved the handling of corrupted control files

Reply:
> Since control files are always contained within a page, that read is atomic

If you assume an underlying`ext2` filesystem, that's probably true (the kernel still decide to interrupt file-system reads). If you run this under NFS (and other network-based filesystems) that's unfortunately not true.

Reply:
In that case we will need to add a checksum in the control file (cc @art-w). At least this PR fixes the issue for some file systems.

Reply:
yea it's great to improve the current situation - it's also nice to be clear about the limitations.

Are the RO instance re-reading the full files on every "reads"?

The only way to get atomic writes in POSIX is via `rename`. I don't have enough context to know if this could work in this case (but we are doing this in various other places of `irmin`, for instance in `ocaml-git` when we update references: https://github.com/mirage/ocaml-git/blob/master/src/git-unix/git_unix.ml#L529-L547

Reply:
> Are the RO instance re-reading the full files on every "sync"?

Yes.

> The only way to get atomic writes in POSIX is via `rename`.

I see, thanks

Reply:
Could you update the comments in the code to reflect that reads might not be atomic - this can be misleading for critical pieces of code

Reply:
Done in https://github.com/mirage/irmin/pull/2107